bee-tangle/src/tangle.rs

```diff
@@ -82,110 +67,54 @@ impl<B: StorageBackend> Tangle<B> {
     }
 
     /// Insert a message into the tangle.
-    pub async fn insert(
-        &self,
-        message: Message,
-        message_id: MessageId,
-        metadata: MessageMetadata,
-    ) -> Option<MessageRef> {
-        let exists = self.pull_message(&message_id, true).await;
-
-        let msg = self.insert_inner(message_id, message.clone(), metadata, !exists).await;
-
-        self.vertices
-            .get_mut(&message_id)
-            .await
-            .expect("Just-inserted message is missing")
-            .allow_eviction();
-
-        if msg.is_some() {
-            // Write parents to DB
-            for &parent in message.parents().iter() {
-                self.storage
-                    .insert(&(parent, message_id), &())
-                    .unwrap_or_else(|e| info!("Failed to update approvers for message {:?}", e));
-            }
-
-            // Insert into backend using hooks
-            self.storage_insert(message_id, message, metadata)
-                .unwrap_or_else(|e| info!("Failed to insert message {:?}", e));
+    pub async fn insert(&self, message: Message, message_id: MessageId, metadata: MessageMetadata) -> Option<Message> {
+        self.storage.insert(&message_id, &message).ok()?;
+        self.storage.insert(&message_id, &metadata).ok()?;
+
+        for &parent in message.parents().iter() {
+            self.storage
+                .insert(&(parent, message_id), &())
+                .unwrap_or_else(|e| info!("Failed to update approvers for message {:?}", e));
         }
 
-        msg
+        Some(message)
```

Comment:
I think this is going to break some stuff. We act on the return value and always assuming something is new may have consequences.

Reply:
I thought this function returns the inserted value (Which is `None` if the insertion failed and `Some` if it was successful). Or is this supposed to return the old value if it was already there?

Reply:
Yeah but I'm not sure insert fails if it already exists...

Reply:
no it does not fail if it already exists

Reply:
I think we don't, we can test this just in case

Reply:
We definitely do, we are using the return value and expect it to be None if it already existed